src/python/pants/jvm/goals/add_artifact_goal.py

```diff
@@ -0,0 +1,232 @@
+) -> CoursierLockfileEntry | None:
+    for entry in lockfile.entries:
+        if entry.coord == coord:
+            return entry
+    return None
+
+
+def _jar_filenames_to_symbols(filenames: Iterable[str]) -> frozenset[str]:
+    symbols: set[str] = set()
+    for filename in filenames:
+        if not filename.endswith(".class"):
+            continue
+        stripped_filename = filename[: -len(".class")]
+        symbols.add(stripped_filename.replace("/", "."))
+    return frozenset(symbols)
+
+
+def _symbols_to_packages(symbols: frozenset[str]) -> tuple[str, ...]:
+    packages: set[str] = set()
+    for symbol in symbols:
+        package, _, _ = symbol.rpartition(".")
+        packages.add(package)
+    return tuple(sorted(frozenset(packages)))
+
+
+@rule
+async def render_jvm_artifact_target_for_build_file(
+    request: RenderJvmArtifactTargetRequest,
+    all_jvm_artifacts: AllJvmArtifactTargets,
+    unzip: UnzipBinary,
+    goal_subsystem: JvmAddArtifactSubsystem,
+) -> RenderJvmArtifactTargetResult:
+    try:
+        coordinate = Coordinate.from_coord_str(request.coordinate_str)
+    except InvalidCoordinateString:
+        return RenderJvmArtifactTargetResult(
+            build_file_content="",
+            build_file_path="",
+            message="This coordinate string is not in the correct format.",
+            is_error=True,
+        )
+
+    # Scan all `jvm_artifact` targets to see if this target is already defined.
+    for tgt in all_jvm_artifacts:
+        req = ArtifactRequirement.from_jvm_artifact_target(tgt)
+        if (
+            req.coordinate.group == coordinate.group
+            and req.coordinate.artifact == coordinate.artifact
+        ):
+            if req.coordinate.version == coordinate.version:
+                message = f"A `jvm_artifact` is already defined at {tgt.address}."
+            else:
+                message = f"A `jvm_artifact` is already defined at {tgt.address}, but has version `{req.coordinate.version}`."
+            return RenderJvmArtifactTargetResult(
+                build_file_content="", build_file_path="", message=message
+            )
+
+    # Resolve this artifact and fetch the applicable jar.
+    lockfile = await Get(
+        CoursierResolvedLockfile, ArtifactRequirements([ArtifactRequirement(coordinate)])
+    )
+    lockfile_entry = _get_entry_for_coord(lockfile, coordinate)
+    if not lockfile_entry:
+        return RenderJvmArtifactTargetResult(
+            build_file_content="",
+            build_file_path="",
+            message=f"Illegal state: Attempted to resolve {coordinate}, but the coordinate was not found in the resolution.",
+            is_error=True,
+        )
+
+    classpath_entry = await Get(ClasspathEntry, CoursierLockfileEntry, lockfile_entry)
+    filenames_result = await Get(
+        ProcessResult,
+        Process(
+            argv=[unzip.path, "-Z", "-1", classpath_entry.filenames[0]],
```

Comment:
I _think_ that this works, but the `ClasspathEntry` for a coursier entry will actually expose the transitive dependencies for the artifact. So you'll want to validate or adjust that code to guarantee that the root/declared artifact is always the first one in the result.

Reply:
Do the filenames follow any consistent naming scheme?

Reply:
They do not. But I believe that this property is already provided: just not currently guaranteed.

Reply:
It seems like the code does put the root entry (the jar itself) first. https://github.com/pantsbuild/pants/blob/960f12f2da2ef2c1d8b0d4fc1bd64336a282ff8d/src/python/pants/jvm/resolve/coursier_fetch.py#L512

However, I also see this TODO nearby:

```
# TODO: Expose an option to control whether this exports only the root, direct dependencies,
# transitive dependencies, etc.
```

Which points to not relying on what appears to be an implementation detail.

Maybe I should follow that TODO's advice and add an option? Where would that option go?